devito/dimension.py

```diff
@@ -533,28 +563,147 @@ def _arg_values(self, *args, **kwargs):
+
+    :param parent: Parent dimension from which the ModuloDimension is created.
+    :param offset: An integer representing an offset from the parent dimension.
+    :param modulo: The extent of the range.
+    :param name: (Optional) force a name for this Dimension.
     """
 
-    def __new__(cls, name, origin, **kwargs):
-        newobj = sympy.Symbol.__new__(cls, name)
-        newobj._origin = origin
+    def __new__(cls, parent, offset, modulo, name=None):
+        return ModuloDimension.__xnew_cached_(cls, parent, offset, modulo, name)
+
+    def __new_stage2__(cls, parent, offset, modulo, name):
+        if name is None:
+            name = cls._genname(parent.name, (offset, modulo))
+        newobj = UnboundedDimension.__xnew__(cls, name, parent)
+        newobj._offset = offset
+        newobj._modulo = modulo
         return newobj
 
+    __xnew_cached_ = staticmethod(cacheit(__new_stage2__))
+
+    @property
+    def offset(self):
+        return self._offset
+
+    @property
+    def modulo(self):
+        return self._modulo
+
     @property
     def origin(self):
-        return self._origin
+        return self.parent + self.offset
 
-    def _hashable_content(self):
-        return Symbol._hashable_content(self) + (self.origin,)
+    @cached_property
+    def symbolic_start(self):
+        return (self.root + self.offset) % self.modulo
+
+    symbolic_incr = symbolic_start
+
+    @property
+    def _properties(self):
+        return (self._offset, self._modulo)
+
+
+class IncrDimension(UnboundedDimension):
+
+    is_UnboundedIncr = True
+
+    """
+    Dimension symbol representing a non-contiguous sub-region of a given
+    ``parent`` Dimension, with one point every ``offset`` points. Thus, if
+    ``offset == k``, the dimension represents the sequence ``start, start + k,
+    start + 2*k, ...``.
+
+    :param parent: Parent dimension from which the IncrDimension is created.
+    :param start: An integer representing the starting point of the sequence.
+    :param offset: The distance between two consecutive points.
+    :param name: (Optional) force a name for this Dimension.
+    """
+
+    def __new__(cls, parent, start, offset, name=None):
+        return IncrDimension.__xnew_cached_(cls, parent, start, offset, name)
+
+    def __new_stage2__(cls, parent, start, offset, name):
+        if name is None:
+            name = cls._genname(parent.name, (start, offset))
+        newobj = UnboundedDimension.__xnew__(cls, name, parent)
+        newobj._start = start
+        newobj._offset = offset
+        return newobj
+
+    __xnew_cached_ = staticmethod(cacheit(__new_stage2__))
+
+    @property
+    def offset(self):
+        return self._offset
+
+    @cached_property
+    def symbolic_start(self):
+        return self._start
+
+    @property
+    def symbolic_incr(self):
+        return self + self.offset
+
+    @property
+    def _properties(self):
+        return (self._start, self._offset)
```

Comment:
Just a comment, but we start to have a looot of different Dimensions type that may be a bit confusing for new users. Not sure if we can merge some.

Reply:
we cannot and we shouldn't merge them. 

However, it's not all so bad. These ``UnboundedDimensions`` are not meant to be used by the user directly. They are rather created internally by Devito. So it shouldn't be a big deal. 

Other dimensions could be hidden as well (think about the discussion on ConditionalDimension the other day on Slack)

Not sure about SubDimensions